File: backend/app/schemas/question.py
"""试题集相关的请求/响应模型"""

from datetime import datetime

from pydantic import BaseModel, Field

from backend.app.models.question_set import QuestionSetStatus


class QuestionGenerateRequest(BaseModel):
    """生成试题集请求"""

    natural_language_request: str = Field(..., min_length=10, max_length=2000)
    course_id: int
    subject: str | None = None
    chapter_id: int | None = None
    difficulty: str | None = Field(default=None, pattern="^(easy|medium|hard)$")


class QuestionModifyRequest(BaseModel):
    """修改试题集请求"""

    natural_language_request: str = Field(..., min_length=5, max_length=2000)


class QuestionSetResponse(BaseModel):
    """试题集响应"""

    id: int
    title: str
    description: str | None
    course_id: int
    teacher_id: int
    is_public: bool
    status: QuestionSetStatus
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class QuestionSetListResponse(BaseModel):
    """试题集列表响应"""

    question_sets: list[QuestionSetResponse]
    total: int


class AssignRequest(BaseModel):
    """分配试题集请求"""

    assign_to_all: bool = False
    student_ids: list[int] | None = None
    deadline: datetime | None = None


class StudentQuestionSetResponse(BaseModel):
    """学生试题集响应"""

    id: int
    title: str
    description: str | None
    is_assigned: bool
    is_completed: bool
    deadline: datetime | None
    completed_at: datetime | None
    has_draft: bool
    course_name: str

    model_config = {"from_attributes": True}


class QuestionSetContentResponse(BaseModel):
    """试题集内容响应"""

    id: int
    title: str
    markdown_content: str

File: frontend/src/services/question.ts
import api from './api';

// 类型定义
export interface QuestionSet {
  id: number;
  title: string;
  description: string | null;
  course_id: number;
  teacher_id: number;
  is_public: boolean;
  status: 'draft' | 'published';
  shared_to_plaza_at: string | null;
  created_at: string;
  updated_at: string;
}

export interface GenerateQuestionRequest {
  course_id: number;
  natural_language_request: string;
  subject?: string | null;
  chapter_id?: number | null;
  difficulty?: 'easy' | 'medium' | 'hard' | null;
}

export interface ModifyQuestionRequest {
  natural_language_request: string;
}

export interface AssignQuestionRequest {
  student_ids: number[];
  deadline?: string; // ISO 8601 格式
  assign_to_all?: boolean;
}

export interface QuestionSetListResponse {
  question_sets: QuestionSet[];
  total: number;
}

export interface StudentQuestionSet {
  id: number;
  title: string;
  description: string | null;
  course_name: string;
  is_assigned: boolean;
  is_completed: boolean;
  deadline: string | null;
  completed_at: string | null;
  has_draft: boolean;
}

export interface StudentQuestionSetListResponse {
  question_sets: StudentQuestionSet[];
  total: number;
}

// 试题集 API（教师端）
export const questionApi = {
  // 生成试题集
  generate: (data: GenerateQuestionRequest) =>
    api.post<QuestionSet>('/question/generate', data),

  // 修改试题集
  modify: (questionSetId: number, data: ModifyQuestionRequest) =>
    api.post<QuestionSet>(`/question/${questionSetId}/modify`, data),

  // 获取试题集内容
  getContent: (questionSetId: number) =>
    api.get<{ id: number; title: string; markdown_content: string }>(`/question/${questionSetId}/content`),

  // 分配试题集
  assign: (questionSetId: number, data: AssignQuestionRequest) =>
    api.post(`/question/${questionSetId}/assign`, data),

  // 获取试题集列表
  list: (courseId: number) =>
    api.get<QuestionSetListResponse>('/question/list', {
      params: { course_id: courseId },
    }),

  // 发布试题集
  publish: (questionSetId: number) =>
    api.post(`/question/${questionSetId}/publish`),

  // 删除试题集
  delete: (questionSetId: number) =>
    api.delete(`/question/${questionSetId}`),
};

// 学生试题集 API
export const studentQuestionApi = {
  // 获取可访问的试题集列表（后端直接返回数组）
  list: (courseId?: number) =>
    api.get<StudentQuestionSet[]>('/student-question/list', {
      params: courseId ? { course_id: courseId } : {},
    }),

  // 获取试题集内容（学生端）
  getContent: (questionSetId: number) =>
    api.get<{ id: number; title: string; markdown_content: string }>(`/student-question/${questionSetId}/content`),
};

